### PR TITLE
Resinify for parallella

### PR DESCRIPTION
**Pull request by pcarranzav**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
+FROM resin/armv7hf-debian:jessie
+
+RUN apt-get update \
+	&& apt-get install -y axel wget htop nano libjpeg-dev libconfig++-dev\
+	libevent-dev git freebsd-glue curl python bison flex\
+	&& apt-get clean \
```
Comment (petrosagg) [inline Dockerfile]: You only need `rm -rf /var/apt/lists/*` here instaed of clean and autoremove

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
+FROM resin/armv7hf-debian:jessie
+
+RUN apt-get update \
+	&& apt-get install -y axel wget htop nano libjpeg-dev libconfig++-dev\
```
Comment (petrosagg) [inline Dockerfile]: small style nit:

``` Dockerfile
RUN apt-get update \
  && apt-get install -y
    axel \
    bison \
    curl \
    flex \
    freebsd-glue \
    git \
    htop \
    libconfig++-dev \
    libevent-dev \
    libjpeg-dev \
    nano \
    python \
    wget
  && rm -rf /var/lib/apt/lists/*
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	libevent-dev git freebsd-glue curl python bison flex\
+	&& apt-get clean \
+        && apt-get autoremove -qqy 
+
+RUN curl -sL https://deb.nodesource.com/setup | bash -
+RUN apt-get install -y build-essential nodejs
```
Comment (petrosagg) [inline Dockerfile]: if you're using nodejs then you can `FROM resin/armv7hf-node:0.10.38` which also includes python, git, curl, build-essential, libjpeg-dev, libevent-dev,

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+RUN mkdir -p /opt/adapteva && cd /opt/adapteva \
+	&& axel -n 10 http://ftp.parallella.org/esdk/beta/esdk.2014.11.20150522_linux_armv7l.tar.gz && tar -xf esdk.2014.11.20150522_linux_armv7l.tar.gz && rm esdk.2014.11.20150522_linux_armv7l.tar.gz \
```
Comment (petrosagg) [inline Dockerfile]: This runs on a build server, I'm not sure axel makes sense here. It also adds up to your image size.

Comment (petrosagg) [inline Dockerfile]: You can do all this without buffering the whole file like so:

``` Dockerfile
RUN mkdir -p /opt/adapteva/esdk \
    && curl -sL http://ftp.parallella.org/esdk/beta/esdk.2014.11.20150522_linux_armv7l.tar.gz | tar xz -C /opt/adapteva/esdk --strip-components=1
```

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	&& ln -sTf /opt/adapteva/esdk.2014.11 /opt/adapteva/esdk
+
+ENV EPIPHANY_HOME /opt/adapteva/esdk
```
Comment (petrosagg) [inline Dockerfile:20]: I'd move that before the above `curl` and use the variable in the command

``` Dockerfile
RUN mkdir -p $EPIPHANY_HOME \
    && curl -sL http://ftp.parallella.org/esdk/beta/esdk.2014.11.20150522_linux_armv7l.tar.gz | tar xz -C $EPIPHANY_HOME --strip-components=1
```

Also, a ESDK_VERSION env var would make sense

``` Dockerfile
ENV EPIPHANY_HOME /opt/adapteva/esdk
ENV ESDK_VERSION 2014.11.20150522

RUN mkdir -p $EPIPHANY_HOME \
    && curl -sL http://ftp.parallella.org/esdk/beta/esdk.$ESDK_VERSION_linux_armv7l.tar.gz | tar xz -C $EPIPHANY_HOME --strip-components=1
```

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+# Enable default setup from webterminal
+RUN sed -i 's/\/bin\/sh/\/bin\/bash/g' /opt/adapteva/esdk/setup.sh && echo "source /opt/adapteva/esdk/setup.sh" >> ~/.bashrc
```
Comment (petrosagg) [inline Dockerfile]: You can use any character as a `sed` separator which can help to minimise escaping. A pipe character for example:

``` bash
sed -i 's|/bin/sh|/bin/bash|g'
```

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+RUN mkdir -p /app
```
Comment (petrosagg) [inline Dockerfile:31]: I suggest moving the source code to `/usr/src/app`

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+# Build libelf 0.8.13
+RUN cd /app && wget http://www.mr511.de/software/libelf-0.8.13.tar.gz \
```
Comment (petrosagg) [inline Dockerfile:34]: This should probably be in `/usr/src/libelf` and again, ENV var for versioning and streaming tar extraction:

``` Dockerfile
ENV LIBELF_VERSION 0.8.13

RUN mkdir -p /usr/src/libelf \
    && curl -sL http://www.mr511.de/software/libelf-$LIBELF_VERSION.tar.gz | tar xz -C /usr/src/libelf --strip-components=1 \
    && cd /usr/src/libelf \
    && ./configure \
    && make \
    && make install
```

I guess you could even delete `/usr/src/libelf` after installation?

File: run.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (petrosagg) [inline run.sh:2]: missing `#!/bin/bash`

File: run.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+if [ ! -f /app/installed.txt ]; then
+	echo "Building FFT correlation"
+	cd /app/parallella-fft-xcorr && PATH=/usr/local/browndeer/bin:$PATH LD_LIBRARY_PATH=/usr/local/browndeer/lib:$LD_LIBRARY_PATH make IMPL=coprthr
+	install -m 644 /app/parallella-fft-xcorr/libfft-demo-coprthr.so /usr/lib/
+	ldconfig /usr/lib /usr/local/browndeer/lib
+
+	cp /app/parallella-fft-xcorr/device.cbin.3.e32 /app/
+	echo "Installing app"
+	cd /app && npm install --unsafe-perm
```
Comment (petrosagg) [inline run.sh:15]: Is there any reason npm modules have to be installed on the device and not during build time?

File: Dockerfile (hunk 2 above)
Comment (petrosagg) [inline Dockerfile]: After the comments bellow, you only need to install `bison flex freebsd-glue libconfig++-dev` and maybe add `htop nano` if you want a debug build

File: Dockerfile (hunk 6 above)
Comment (petrosagg) [inline Dockerfile]: Are you sure this doesn't work as it is?

EDIT: It can work without the `sed`, see bellow

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	&& tar -xf libelf-0.8.13.tar.gz \
+	&& cd libelf-0.8.13 && ./configure && make && make install
+
+# Build COPRTHR
+RUN bash -c "source /opt/adapteva/esdk/setup.sh && cd /app \
```
Comment (petrosagg) [inline Dockerfile:39]: Same drill. Also you don't need to wrap everything in a string and run via `/bin/bash`

``` Dockerfile
ENV COPRTHR_VERSION parallellocalypse

RUN . /opt/adapteva/esdk/setup.sh \
    && mkdir -p /usr/src/libcoprthr \
    && curl -sL https://github.com/olajep/coprthr/archive/$COPRTHR_VERSION.tar.gz | tar xz -C /usr/src/libcoprthr --strip-components=1 \
    && cd /usr/src/libcoprthr \
    && ./configure --enable-epiphany \
    && make \
    && make install
```

I guess you could even delete `/usr/src/libcoprthr` after installation?

File: run.sh (hunk 10 above)
Comment (pcarranzav) [inline run.sh:15]: Not really. I got to this after fighting a while with node-gyp and incompatible v8 versions. But I'll do the npm install at build time and just run `node-gyp configure build here`.

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	&& git clone https://github.com/olajep/coprthr.git \
+	&& cd /app/coprthr && git checkout parallellocalypse && ./configure --enable-epiphany \
+	&& make && make install"
+
+# Add COPRTHR MPI
```
Comment (petrosagg) [inline Dockerfile:45]:
``` Dockerfile
ENV LIBCOPTHR_MPI_VERSION preview

RUN . /opt/adapteva/esdk/setup.sh \
    && mkdir -p /usr/src/libcoprthr-mpi \
    && curl -sL http://www.browndeertechnology.com/code/bdt-libcoprthr_mpi-$LIBCOPTHR_MPI_VERSION.tgz | tar xz -C /usr/src/libcoprthr-mpi \
    && cd /usr/src/libcoprthr-mpi \
    && ./install.sh
```

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+RUN bash -c "source /opt/adapteva/esdk/setup.sh && cd /app && wget http://www.browndeertechnology.com/code/bdt-libcoprthr_mpi-preview.tgz \
+	&& tar -xf bdt-libcoprthr_mpi-preview.tgz \
+	&& cd /app/libcoprthr_mpi && ./install.sh"
+
+# Clone the FFT correlation repo
+RUN cd /app && git clone https://github.com/olajep/parallella-fft-xcorr \
+	&& cd /app/parallella-fft-xcorr && git checkout c2bee839535bcff868cdeb7c1c5f735a60d02f44
+
+ADD . /app
```
Comment (petrosagg) [inline Dockerfile:57]: Use `COPY` instead of `ADD`. Also `COPY . /usr/src/app`

File: Dockerfile [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	&& cd /app/coprthr && git checkout parallellocalypse && ./configure --enable-epiphany \
+	&& make && make install"
+
+# Add COPRTHR MPI
+RUN bash -c "source /opt/adapteva/esdk/setup.sh && cd /app && wget http://www.browndeertechnology.com/code/bdt-libcoprthr_mpi-preview.tgz \
+	&& tar -xf bdt-libcoprthr_mpi-preview.tgz \
+	&& cd /app/libcoprthr_mpi && ./install.sh"
+
+# Clone the FFT correlation repo
+RUN cd /app && git clone https://github.com/olajep/parallella-fft-xcorr \
```
Comment (petrosagg) [inline Dockerfile]: and once more

``` Dockerfile
ENV PARALLELLA_FFT_XCORR_VERSION c2bee839535bcff868cdeb7c1c5f735a60d02f44

RUN curl -sL https://github.com/olajep/parallella-fft-xcorr/archive/$PARALLELLA_FFT_XCORR_VERSION.tar.gz | tar xz -C /usr/src/app --strip-components=1
```

File: run.sh [hunk 15]
```diff
@@ -0,0 +1,19 @@
+
+set -e
+
+source /opt/adapteva/esdk/setup.sh
+
+if [ ! -f /app/installed.txt ]; then
+	echo "Building FFT correlation"
```
Comment (petrosagg) [inline run.sh:8]: is there any reason this can't be built on the build server?

Comment (pcarranzav) [inline run.sh:8]: Yes. The CL linker from COPRTHR looks for /dev/epiphany at build time, so it fails on the build server.

File: run.sh (hunk 10 above)
Comment (petrosagg) [inline run.sh:15]: Why can't the whole thing be done in the build server?

File: run.sh (hunk 15 above)
Comment (petrosagg) [inline run.sh:8]: hm, ok.

**Comment by petrosagg** [comment]
superseded by pull #4

**Pull request closed without merging**